test-data/unit/check-overloading.test

```diff
@@ -17,6 +17,121 @@ class A: pass
 class B: pass
 [builtins fixtures/isinstance.pyi]
 
+[case testOverloadNeedsImplementation]
+from typing import overload, Any
+@overload  # E: An overloaded function outside a stub file must have an implementation
+def f(x: 'A') -> 'B': ...
+@overload
+def f(x: 'B') -> 'A': ...
+
+reveal_type(f(A())) # E: Revealed type is '__main__.B'
+reveal_type(f(B())) # E: Revealed type is '__main__.A'
+
+class A: pass
+class B: pass
+[builtins fixtures/isinstance.pyi]
+
+[case testSingleOverloadNoImplementation]
+from typing import overload, Any
+@overload  # E: Single overload definition, multiple required
+def f(x: 'A') -> 'B': ...
+
+class A: pass
+class B: pass
+[builtins fixtures/isinstance.pyi]
+
+
+[case testOverloadByAnyOtherName]
+from typing import overload as rose
+from typing import Any
+@rose
+def f(x: 'A') -> 'B': ...
+@rose
+def f(x: 'B') -> 'A': ...
+
+def f(x: Any) -> Any:
+    pass
+
+reveal_type(f(A())) # E: Revealed type is '__main__.B'
+reveal_type(f(B())) # E: Revealed type is '__main__.A'
+
+class A: pass
+class B: pass
+[builtins fixtures/isinstance.pyi]
+
+[case testTypeCheckOverloadWithDecoratedImplementation]
```

Comment:
How about another test that checks for errors where the implementation doesn't match the overloads.